### PR TITLE
Proposal: add option to apply changes/elab on save

### PR DESCRIPTION
**Pull request by ammkrn**
This is a nice option to have in general, but the primary motivation is that elaborating on every change doesn't work well with vscode if users have an input extension enabled like the vim one which is mad popular, or dance (real heads know) and are working on a file larger than a few hundred lines. You end up with a pretty heavy amount of lag/eaten inputs, and when you start doing the multiple cursor stuff shit gets real crazy.
I did a little bit of reading about this a while ago and it has something to do with vscode's extension architecture and how it handles stuff that reads inputs, so I don't think it's the fault of mm0-rs.

The basic idea is that when the server is set to trigger on save instead of change, `didChange` messages have their deltas pushed to a queue, and the next `didSave` or `didClose` message will apply all of the queued changes in the order they were received.

This just adds the feature and a CLI flag for it, but currently the only way to turn it on is to edit the source and recompile (with `ServerOptions.elab_on set to Some(Save)`). There's an auxiliary change that adds `Mutex<ServerOptions>` to the `Server` struct which may or may not be desirable going forward.

PS: I think you might have to update the extension on the vscode marketplace since it's still showing up as "Example configuration" instead of "mm0" or however you want to style it.

### CODE REVIEW

**Comment by digama0** [comment]
On a `didClose` message, where do we post the diagnostics? Currently we treat that only as a "revert to saved" modification, if there are dependent files, and drop the file without elaboration otherwise.

**Comment by digama0** [comment]
Also: CI works! This is great.

File: mm0-rs/src/server.rs [hunk 1]
```diff
@@ -1211,26 +1305,45 @@ impl Server {
...
+                if options.ulock().elab_on == Some(ElabOn::Save) {
+                  apply_queued_changes(vfs, &change_queues, path.clone(), None)?;
+                }
                 vfs.close(&path)?;
+              },
+              "textDocument/didSave" => {
+                  let DidSaveTextDocumentParams { text_document: doc, .. } = from_value(notif.params)?;
+                  let path = FileRef::from(doc.uri);
+                  log!("save {:?}", path);
+                  if options.ulock().elab_on == Some(ElabOn::Save) {
```
Comment (digama0) [inline mm0-rs/src/server.rs:1344]:
```suggestion
                  if options.ulock().elab_on.or_default() == ElabOn::Save {
```

**Comment by digama0** [comment]
Does this require changes to the vscode extension to enable? I think we need a setting for "additional command line arguments" to mm0-rs so that folks can specify this.

Alternatively, we could add commands to the metaprogramming language to set options on the server. That's a bit weird since there is no "go" button when it comes to elaboration so you could end up setting a server variable to an undesirable value as you type, but the advantage is that it keeps the interface unified.

**Comment by ammkrn** [comment]
> On a `didClose` message, where do we post the diagnostics? Currently we treat that only as a "revert to saved" modification, if there are dependent files, and drop the file without elaboration otherwise.

Actually it might be best to revert `didClose` entirely. The thing I had in mind was someone working without any kind of auto save, but if that's the case, then the two behaviors would be:
1. a user closes the file, but auto save triggers a didSave message which properly updates the VFS (in which case I guess you should just log any errors to the vscode output)
2. a user closes the file, but with no auto-save, so you don't actually want the changes to be reflected in the VFS since the next time it's opened, it should be in the pre-save state. 

In case 2, it might be best to have one or both of `onClose` and `onOpen` (to account for a tab crashing or something, idk) clear any queued changes that weren't claimed directly by a `didSave` message.

**Comment by ammkrn** [comment]
> Does this require changes to the vscode extension to enable? I think we need a setting for "additional command line arguments" to mm0-rs so that folks can specify this.

IMO the most intuitive option would be to have it appear as a checkbox/toggle thing in the vscode extension settings for mm0, which I think requires changing the package.json at least. When you change the values for the options that are already there (like the max number of problems thing), the mm0-rs server loop picks up a message from that event, so it might be possible to pick it up entirely through that.

**Comment by ammkrn** [comment]
Actually that's a bad commit, I need to see what happens if the same file is being worked on in two different windows, my bad.

**Comment by ammkrn** [comment]
The multiple windows thing is pretty annoying. If a user opens `peano.mm1` in two different windows, makes changes in window 1, more changes in window 2, and then closes window 1 without saving, there's no simple way to remove only the deltas from window 1 from the queue.

A workaround I'm using right now is to register this server capability:
```
    regs.push(Registration {
      id: String::new(),
      method: "textDocument/didSave".into(),
      register_options: Some(to_value(
        lsp_types::SaveOptions { include_text: Some(true) }
      ).expect("failed to didSave registration"))
    });
```

to get the whole (new) text file back in `didSave` as part of `DidSaveTextDocumentParams` then just push that to the vfs, bump the version, and call elaborate_and_report with position = None. I'm not sure whether this has any larger negative implications aside from the size of the messages being sent, since they now contain the whole file. The timer in the messages window seems to show about the same duration for elaboration as it does for elaborating on change.

I'm going to close this for now though.

**Comment by digama0** [comment]
> The multiple windows thing is pretty annoying. If a user opens peano.mm1 in two different windows, makes changes in window 1, more changes in window 2, and then closes window 1 without saving, there's no simple way to remove only the deltas from window 1 from the queue.

I haven't really handled multiple windows so far. I know there is a bug where you can open the file twice and close it twice and the server gets confused because it expects only one close command, but I haven't looked into it much.

Rather than queuing up changes, you can just apply all changes eagerly to the file, and just hold off on elaboration until save time (although other actions like a hover request will also trigger elaboration, not sure whether that should be suppressed).

You also don't need to remove some deltas and not others from the queue. If all changes are applied as they arrive, then you always have the correct file data. If the file is completely closed in all windows, then you reload the file from disk if you still need it around; if this happens post-save, this will be the updated data, and if the user didn't save then this will revert to the on disk version.

**Comment by ammkrn** [comment]
> You also don't need to remove some deltas and not others from the queue. If all changes are applied as they arrive, then you always have the correct file data. If the file is completely closed in all windows, then you reload the file from disk if you still need it around; if this happens post-save, this will be the updated data, and if the user didn't save then this will revert to the on disk version.

That's a good idea. It leaves the edge case weirdness where closing window 2 without saving, then saving window 1 will reflect the changes from both in window 1, but yeah with the eager application it will fix itself once both windows are closed and the file is re-opened.

Of the other elaboration triggers, hover seems like the only one that demands elaboration with the absolute newest environment since it needs an the spans to match up with what the user is seeing, but the bright side is that hover isn't going to trigger while people are typing. The document symbol and completion request ones are triggered much more often (and during typing), but unless someone's making a major refactor it might not be a huge deal if it shows stuff from the last saved environment.

**Comment by digama0** [comment]
> That's a good idea. It leaves the edge case weirdness where closing window 2 without saving, then saving window 1 will reflect the changes from both in window 1,

Isn't that the way vscode works? If you open the same file in two windows, they still have the same backing text, changes in one are immediately reflected in the other.

> The document symbol and completion request ones are triggered much more often (and during typing), but unless someone's making a major refactor it might not be a huge deal if it shows stuff from the last saved environment.

The completion request already uses the last saved environment. Document symbol is actually kind of annoying, I've considered removing it because it is pretty big and vscode asks for it way too often. We can also approximate it by applying the text changes to the spans in the document symbol request, only updating it when we need to elaborate for some other reason.

**Comment by ammkrn** [comment]
> Isn't that the way vscode works? If you open the same file in two windows, they still have the same backing text, changes in one are immediately reflected in the other.

If you have auto save off or set to a delay, the changes are only reflected in one window until the file is saved, at which point they both get updated. If the save reveals a conflict it will give you a pop-up asking which version you want to keep.

> The completion request already uses the last saved environment. 

Ah I see it now `let (text, env) = if let Some(old) = old { old } else { ...`, good to know.

**Comment by digama0** [comment]
> If you have auto save off or set to a delay, the changes are only reflected in one window until the file is saved, at which point they both get updated. If the save reveals a conflict it will give you a pop-up asking which version you want to keep.

Oh, that's news to me. If I use ctrl-drag or similar to open a file twice in split-screen, edits in one view are reflected in the other. This is without any saving going on, just in-memory editor panes. If it's possible to open a file twice such that they can be saved independently, then that sort of breaks the editor model, because we represent a file on disk only once. If you have file A open twice with two different contents, and file B depends on A, which version should it use?

**Comment by ammkrn** [comment]
> Oh, that's news to me. If I use ctrl-drag or similar to open a file twice in split-screen, edits in one view are reflected in the other. This is without any saving going on, just in-memory editor panes. If it's possible to open a file twice such that they can be saved independently, then that sort of breaks the editor model, because we represent a file on disk only once. If you have file A open twice with two different contents, and file B depends on A, which version should it use?

If by split-screen you mean tabs then yeah, even if you have auto save turned off, tabs in the same window share the same contents and are updated automatically. I'm not sure why they made windows behave differently.

It looks like what they do for opened but not saved files is (for each actual window) make a new file with a header that has some metadata and the file's contents (on linux they're in ~/.config/Code/Backups), let you edit that, and then "commit" the new contents to the 'real' file when you go to save. The headers have timestamps and the URI. So I guess their way of solving it was to make it so that editing files without auto save on is actually editing a separate buffer, or multiple separate buffers, but when you go to save you have to resolve any conflicts that arise.

I'm not sure how to get the relevant information through the server, but the RLS guys seem to have figured this (or some other workaround) out since it can catch syntax errors in unsaved files. I check with them tomorrow or something and see if they have any sage advice.

**Comment by bjorn3** [comment]
Both RLS and rust-analyzer have a VFS internally of which they base all action on. RLS uses rustc as library to present it that VFS as if it was the real file system. Rust-analyzer requires the user to save files for `cargo check` to run and report errors as using rustc as library is unstable. It also reports certain kinds of errors itself based on the VFS.

**Comment by ammkrn** [comment]
> Both RLS and rust-analyzer have a VFS internally of which they base all action on. RLS uses rustc as library to present it that VFS as if it was the real file system. Rust-analyzer requires the user to save files for `cargo check` to run and report errors as using rustc as library is unstable. It also reports certain kinds of errors itself based on the VFS.

Thanks, I meant rust-analyzer, my mistake. The thing is that there seems to be one more layer of indirection; if you don't have auto save enabled, the changes you're making in the editor aren't made to the actual file, they're made to an intermediate file stored by vscode whose location is OS-specific and not reflected in LSP messages, yet RA seems to know how to find/read it. If I make a rust file:

```
fn main() {
  println!("Hello world");
}
```

and make a change in the editor but don't save it:
```
fn main() {
  !!
}
```

the contents of the actual hello_world.rs are unchanged, but there's another file created at this location:
`/Users/ammkrn_/Library/Application Support/Code/Backups/a48ddb4c71bdfadb578b9b579e4c5615/file/c613dfeeb65c7b253e3d67df53cc5ae4`
with the contents:
```
file:///Users/ammkrn_/rust/demo/src/hello_world.rs {"mtime":1606037404688,"ctime":1606037195502,"size":0,"etag":"3630kbsij0","orphaned":false}

fn main() {
    !!
}
```

and RA seems to be able to see it, since RA returns the error `Syntax Error: expected expressionrust-analyzer` with the correct span highlighted. The workspace option `"rust-analyzer.trace.server": "verbose"` doesn't log anything that mentions this intermediate file.

**Comment by bjorn3** [comment]
> The thing is that there seems to be one more layer of indirection; if you don't have auto save enabled, the changes you're making in the editor aren't made to the actual file, they're made to an intermediate file stored by vscode whose location is OS-specific and not reflected in LSP messages, yet RA seems to know how to find/read it.

No, it doesn't. It applies the `didChange` messages to it's internal VFS. The `didChange` messages keep the vscode and rust-analyzer file state in sync without rust-analyzer directly reading the state. In fact there have been bugs in the past where both states got out of sync. For example with non-ascii characters. (This happened because lsp uses UTF-16 offsets, while rust-analyzer uses UTF-8 offsets. There were bugs in the translation between the two)

**Comment by ammkrn** [comment]
> No, it doesn't. It applies the `didChange` messages to it's internal VFS. The `didChange` messages keep the vscode and rust-analyzer file state in sync without rust-analyzer directly reading the state. 

Thanks for making that clear. I wasn't sure why two open instances of the same file weren't causing issues, but the insight I was missing is that if you have two vscode windows open they have two separate instances of the server running, so they have their own vfs (at least for mm0).

**Pull request closed without merging**